packages/router/router.ts

```diff
@@ -457,6 +521,18 @@ type FetcherStates<TData = any> = {
 export type Fetcher<TData = any> =
   FetcherStates<TData>[keyof FetcherStates<TData>];
 
+export type Blocker =
+  | { state: "blocked"; reset(): void; proceed(): Promise<void>; fn(): boolean }
+  | { state: "unblocked"; reset: undefined; proceed: undefined; fn(): boolean }
+  | {
+      state: "proceeding";
+      reset: undefined;
+      proceed: undefined;
+      fn(): boolean;
+    };
+
+export type ShouldBlockFunction = () => boolean;
```

Comment:
is it a big effort to support this in async as well?

Reply:
Clicking a link or the back button triggers a navigation immediately, so the blocker would never work if we had to wait on some async task before we even know whether or not we're supposed to block.